// @ant-design/pro-layout/lib/SettingDrawer/ThemeColor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/lib/tooltip/style");

var _tooltip = _interopRequireDefault(require("antd/lib/tooltip"));

require("antd/lib/icon/style");

var _icon = _interopRequireDefault(require("antd/lib/icon"));

require("./ThemeColor.less");

var _react = _interopRequireDefault(require("react"));

var _utils = require("../utils/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var Tag = function Tag(_a) {
  var color = _a.color,
      check = _a.check,
      rest = __rest(_a, ["color", "check"]);

  return _react.default.createElement("div", Object.assign({}, rest, {
    style: {
      backgroundColor: color
    }
  }), check ? _react.default.createElement(_icon.default, {
    type: "check"
  }) : '');
};

var ThemeColor = function ThemeColor(_ref) {
  var colors = _ref.colors,
      title = _ref.title,
      value = _ref.value,
      onChange = _ref.onChange,
      formatMessage = _ref.formatMessage;
  var colorList = colors || [];

  if (colorList.length < 1) {
    return null;
  }

  return _react.default.createElement("div", {
    className: "theme-color"
  }, _react.default.createElement("h3", {
    className: "theme-color-title"
  }, title), _react.default.createElement("div", {
    className: "theme-color-content"
  }, colorList.map(function (_ref2) {
    var key = _ref2.key,
        color = _ref2.color;
    var themeKey = (0, _utils.genThemeToString)(key);
    return _react.default.createElement(_tooltip.default, {
      key: color,
      title: themeKey ? formatMessage({
        id: "app.setting.themecolor.".concat(themeKey)
      }) : key
    }, _react.default.createElement(Tag, {
      className: "theme-color-block",
      color: color,
      check: value === key || (0, _utils.genThemeToString)(value) === key,
      onClick: function onClick() {
        return onChange && onChange(key);
      }
    }));
  })));
};

var _default = ThemeColor;
exports.default = _default;
