// @ant-design/pro-layout/es/SettingDrawer/ThemeColor.js
import "antd/es/tooltip/style";
import _Tooltip from "antd/es/tooltip";
import "antd/es/icon/style";
import _Icon from "antd/es/icon";

var __rest = this && this.__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

import './ThemeColor.less';
import React from 'react';
import { genThemeToString } from '../utils/utils';

var Tag = function Tag(_a) {
  var color = _a.color,
      check = _a.check,
      rest = __rest(_a, ["color", "check"]);

  return React.createElement("div", Object.assign({}, rest, {
    style: {
      backgroundColor: color
    }
  }), check ? React.createElement(_Icon, {
    type: "check"
  }) : '');
};

var ThemeColor = function ThemeColor(_ref) {
  var colors = _ref.colors,
      title = _ref.title,
      value = _ref.value,
      onChange = _ref.onChange,
      formatMessage = _ref.formatMessage;
  var colorList = colors || [];

  if (colorList.length < 1) {
    return null;
  }

  return React.createElement("div", {
    className: "theme-color"
  }, React.createElement("h3", {
    className: "theme-color-title"
  }, title), React.createElement("div", {
    className: "theme-color-content"
  }, colorList.map(function (_ref2) {
    var key = _ref2.key,
        color = _ref2.color;
    var themeKey = genThemeToString(key);
    return React.createElement(_Tooltip, {
      key: color,
      title: themeKey ? formatMessage({
        id: "app.setting.themecolor.".concat(themeKey)
      }) : key
    }, React.createElement(Tag, {
      className: "theme-color-block",
      color: color,
      check: value === key || genThemeToString(value) === key,
      onClick: function onClick() {
        return onChange && onChange(key);
      }
    }));
  })));
};

export default ThemeColor;
